### ISSUE TITLE
require-jsdoc with different reporting levels

### ISSUE THREAD

**Issue by marmaak** [gajus/eslint-plugin-jsdoc#862]
## Motivation

Currently if I'm not mistaken the `require-jsdoc` rule can be used with a single kind of reporting level level (error or warning), but it might be useful to have different levels based on the `contexts` specified (I'm even not sure this is feasible from an eslint engine perspective).

An example can be:

```json
{
  "rules": {
    "jsdoc/require-jsdoc": [
      "error",
      {
        "contexts": [
          "MethodDefinition",
          { "MethodDefinition:not[accessibility=\"public\"]": "warn" }
        ]
      }
    ]
  }
}
```
With the above configuration ESLint is supposed to report as an `"error"` all `public` `MethodDefinition`s where JSDocs are missing but only `"warn"` about those occurrences where the accessibility is `protected` or `private`. 

## Current behavior

The reporting level can be set only to `"error"` or `"warn"`

## Desired behavior

There should be the possibility to set a more granular reporting level

## Alternatives considered

I did some researches, and the only suggestion I found is this one

https://stackoverflow.com/questions/34077756/including-both-errors-and-warnings-for-same-eslint-rule

But honestly I don't like it that much 😅 
N/A

**Comment by brettz9** [comment]
ESLint only provides the two levels, and they can only be applied at the whole-rule level.

Creating separate instances of the rule is, as far as I can see, your only other option to truly get such a capability.

However, if you don't mind the fact that there would be no fixers for the contexts you chose to move to this rule (and if you don't have any conflicts by already using the rule) and it is using an experimental parser, you could use our ESLint's built-in `no-restricted-syntax` rule with `@es-joy/jsdoc-eslint-parser` and config something like:

```js
  "parser": "@es-joy/jsdoc-eslint-parser/typescript",

"rules": {
    "no-restricted-syntax": [
      "error",
      {
        "selector": "MethodDefinition[accessibility=\"public\"]:not(:has(JsdocBlock))"
      }
    ],
```

A warning that this parser is experimental (though its `typescript` parser passes off most of the work to `@typescript-eslint/parser`).

Note that you cannot use our version of `jsdoc/no-restricted-syntax` which is geared *not* for finding comment AST (or its lack) but finding AST that has a comment attached.

**Comment by brettz9** [comment]
And the reason our code is not geared for finding comment AST is because it has typically been excluded by parsers from having its own AST. The experimental parser I mentioned, adds it as a more first class citizen, thus allowing checks which are more complex (such as allowing querying both the JavaScript structure and the comment AST in the same expression) than just additively looking for some JavaScript AST and then filtering that by comment AST as our `no-restricted-syntax` and `no-missing-syntax` rules do.

ESLint's own rule is thus more suitable in this case, even though it deals with our specialty of comments, when using proper AST for both JS and JSDoc together, one can treat comments more like other syntax, and thus ESLint's generic rule is more suitable (though unfortunately ESLint doesn't have an equivalent to our `no-missing-syntax` rule).

**Comment by marmaak** [comment]
Thanks for the detailed explanation and also for all suggestions. I think we can close this issue since there is no way currently to improve the plugin since it's mostly related to a limitation (of course I'm not questioning if it can be a right architectural decision or not) of ESLint.

I will keep an eye on that parser and maybe in future I will give it a try, as of now I neglect the idea of differentiating the reporting severity based of the accessibility of class fields
